roles/openshift_health_checker/action_plugins/openshift_health_check.py

```diff
@@ -201,12 +215,131 @@ def run_check(name, check, user_disabled_checks):
+    if check.changed:
+        result["changed"] = True
+    if check.failures or result.get("failed"):
+        if result.get("msg"):  # failure result has msg; combine with any registered failures
+            check.register_failure(result.get("msg"))
+        result["failures"] = [(fail.name, str(fail)) for fail in check.failures]
+        result["failed"] = True
+        result["msg"] = "\n".join(str(fail) for fail in check.failures)
+        write_to_output_file(output_dir, name + ".failures.json", result["failures"])
+    if check.notes:
+        write_to_output_file(output_dir, name + ".notes.json", check.notes)
+    if output_dir and check.files_to_save:
+        write_files_to_save(check, os.path.join(output_dir, name))
+
+    return result
+
+
+def prepare_output_dir(dirname):
+    """Create the directory, including parents. Return bool for success/failure."""
+    try:
+        os.makedirs(dirname)
+        return True
+    except OSError as exc:
+        # trying to create existing dir leads to error;
+        # that error is fine, but for any other, assume the dir is not there
+        return exc.errno == errno.EEXIST
+
+
+def copy_remote_file_to_dir(check, file_to_save, output_dir, fname):
+    """Copy file from remote host to local file in output_dir, if given."""
+    if not output_dir or not prepare_output_dir(output_dir):
+        return
+    local_file = os.path.join(output_dir, fname)
+
+    # pylint: disable=broad-except; do not need to do anything about failure to write dir/file
+    # and do not want exceptions to break anything.
+    try:
+        # NOTE: it would have been nice to copy the file directly without loading it into
+        # memory, but there does not seem to be a good way to do this via ansible.
+        result = check.execute_module("slurp", dict(src=file_to_save), register=False)
+        if result.get("failed"):
+            display.warning("Could not retrieve file {}: {}".format(file_to_save, result.get("msg")))
+            return
+
+        content = result["content"]
+        if result["encoding"] == "base64":
+            content = base64.b64decode(content)
+        with open(local_file, "wb") as outfile:
+            outfile.write(content)
+    except Exception as exc:
+        display.warning("Failed writing remote {} to local {}: {}".format(file_to_save, local_file, exc))
+        return
+
+
+def _no_fail(obj):
+    # pylint: disable=broad-except; do not want serialization to fail for any reason
+    try:
+        return str(obj)
+    except Exception:
+        return "[not serializable]"
+
+
+def write_to_output_file(output_dir, filename, data):
+    """If output_dir provided, write data to file. Serialize as JSON if data is not a string."""
+
+    if not output_dir or not prepare_output_dir(output_dir):
+        return
+    filename = os.path.join(output_dir, filename)
+    try:
+        with open(filename, 'w') as outfile:
+            if isinstance(data, basestring):
```

Comment:
python3 doesn't like basestring. Should I be doing:

    from past.builtins import basestring 

...or something else?

Reply:
https://github.com/openshift/openshift-ansible/blob/20cf3396e4aae5aef6485a23e6cc3134c7a2d1be/roles/openshift_health_checker/action_plugins/openshift_health_check.py#L10
https://github.com/openshift/openshift-ansible/blob/20cf3396e4aae5aef6485a23e6cc3134c7a2d1be/roles/openshift_health_checker/action_plugins/openshift_health_check.py#L183

Reply:
thanks. apparently i'm blind, there being already an example in the file.